### PR TITLE
Actually use the IdStrategy

### PR DESCRIPTION
**Pull request by stephenc**
@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
:bee:

File: src/test/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealmTest.java [hunk 1]
```diff
@@ -45,4 +47,14 @@
...
+        assertEquals("[alice, debbie]", r.loadGroupByGroupname("ADMIN", true).getMembers().toString());
+        assertEquals("[bob]", r.loadGroupByGroupname("dEv", true).getMembers().toString());
+        assertEquals("[charlie, debbie]", r.loadGroupByGroupname("qA", true).getMembers().toString());
+    }
+
+    @Test public void getUserWithIdStrategy() {
+        assertThat(r.loadUserByUsername("alice").getUsername(), is(r.loadUserByUsername("Alice").getUsername()));
+    }
+
+}
```
Comment (tfennelly) [inline src/test/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealmTest.java]: A test for `UsernameNotFoundException` would also be nice to see?

**Comment by tfennelly** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealm.java [hunk 2]
```diff
@@ -122,7 +122,7 @@ protected Random initialValue() {
     @Override protected UserDetails authenticate(String username, String password) throws AuthenticationException {
         doDelay();
         UserDetails u = loadUserByUsername(username);
-        if (!password.equals(username)) {
+        if (!password.equals(u.getUsername())) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealm.java]: Would not be a bad idea to verify that this actually works. Could probably call

```java
r.createSecurityComponents().manager.authenticate(new UsernamePasswordAuthenticationToken("Alice", "alice"));
```

File: src/main/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealm.java [hunk 3]
```diff
@@ -130,37 +130,48 @@ protected Random initialValue() {
...
+                if (idStrategy.equals(group, groupname)) {
+                    return new GroupDetails() {
+                        @Override
+                        public String getName() {
+                            return group;
+                        }
+
+                        @Override
+                        public Set<String> getMembers() {
+                            final IdStrategy idStrategy = getGroupIdStrategy();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealm.java]: :ant: already defined, no need to override.

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealm.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/mocksecurityrealm/MockSecurityRealm.java]: may be changed between when the group is retrieved and stored in a http session and when it was retrieved. hence the on-demand lookup

**Pull request merged**